### PR TITLE
[ONNX] Fix linalg norm output's shapes and dtypes

### PR DESCRIPTION
**Pull request by qqaatw**
Part of #79263

This PR fixes the following matters:

1. Before this fix, the reduced output has `[1]` shape when `dim = None` and `keepdim = False`. Now the output is reduced to `[]` shape, which matches Pytorch's behavior.
2. Before this fix, the output is always casted to `Long`. Now the output is casted to the input's dtype.

### CODE REVIEW

File: torch/onnx/symbolic_opset9.py [hunk 1]
```diff
@@ -4298,7 +4298,9 @@ def linalg_vector_norm(g, self, ord, dim, keepdim, dtype):
     # Conditions based on https://pytorch.org/docs/stable/generated/torch.linalg.vector_norm.html
     if dim is None:
         self = symbolic_helper._reshape_helper(g, self, [-1])
-        keepdim = None
+        keepdim = 0
+    else:
+        keepdim = int(keepdim)
```
Comment (justinchuby) [inline torch/onnx/symbolic_opset9.py:4303]: Q: Why is this needed?

File: torch/onnx/symbolic_opset9.py [hunk 2]
```diff
@@ -4298,7 +4298,9 @@ def linalg_vector_norm(g, self, ord, dim, keepdim, dtype):
     # Conditions based on https://pytorch.org/docs/stable/generated/torch.linalg.vector_norm.html
```
Comment (justinchuby) [inline torch/onnx/symbolic_opset9.py:4298]: Could you help us annotate the function parameter types to make them clear too? Thanks a lot!

**Review by justinchuby — COMMENTED** [review]
Thanks!

**Review by justinchuby — APPROVED** [review]
LGTM. nit lint errors

File: torch/onnx/symbolic_opset9.py (hunk 1 above)
Comment (qqaatw) [reply torch/onnx/symbolic_opset9.py:4303]: In case of `keepdim` is `None`, we cast it to `0`. ONNX seems to treat `None` and `0` as different.

File: torch/onnx/symbolic_opset9.py (hunk 2 above)
Comment (qqaatw) [reply torch/onnx/symbolic_opset9.py:4298]: No prob.

File: torch/onnx/symbolic_opset11.py [hunk 3]
```diff
@@ -1062,12 +1062,17 @@ def linalg_vector_norm(g, self, ord, dim, keepdim, dtype):
             self = symbolic_helper._reshape_helper(
                 g, self, g.op("Constant", value_t=torch.tensor([-1], dtype=torch.int64))
             )
-            keepdim = None
+            keepdim = 0
+        else:
+            keepdim = int(keepdim)
```
Comment (justinchuby) [inline torch/onnx/symbolic_opset11.py:1067]: Is casting needed here too?

File: torch/onnx/symbolic_opset11.py [hunk 4]
```diff
@@ -1062,12 +1062,15 @@ def linalg_vector_norm(g, self, ord, dim, keepdim, dtype):
...
+            keepdim = 0
+
         cond_op = g.op(
             "Not", g.op("Equal", self, g.op("Constant", value_t=torch.LongTensor([0])))
         )
         cond_op = g.op(
-            "Cast", cond_op, to_i=symbolic_helper.cast_pytorch_to_onnx["Long"]
+            "Cast",
+            cond_op,
+            to_i=symbolic_helper.cast_pytorch_to_onnx[self.type().scalarType()],
```
Comment (justinchuby) [inline torch/onnx/symbolic_opset11.py:1073]: This needs to be long because it is a accumulator

Comment (justinchuby) [reply torch/onnx/symbolic_opset11.py:1073]: Also prefer torch.int64 for clarity here and elsewhere

Comment (qqaatw) [reply torch/onnx/symbolic_opset11.py:1073]: > This needs to be long because it is a accumulator

1. From the PyTorch impl. it isn't limited to long and depends on the input scalar type.
2. Why should it be long as an accumulator? I did't hear this account before.
3. `ReduceSum` op allows a variety of dtypes (integers and floating points).
4. Maybe there is a better name suitable than `cond_op`.

Comment (justinchuby) [reply torch/onnx/symbolic_opset11.py:1073]: Sorry I confused this w/ another pr. SGTM

**Review by BowenBao — APPROVED** [review]
🚀

**Comment by justinchuby** [comment]
@pytorchbot merge -g

may need to rebase
